### DISCUSSION TITLE
How to use multiple runtimes for serverless functions WITHOUT breaking local dev

### DISCUSSION THREAD

**Discussion by slimshreydy** [vercel/vercel#9840] [General]
### Description

I'm interested in including a mix of Python + Nextjs/node-based functions in my deployed API. I've seen https://github.com/vercel/vercel/discussions/6197, which suggests that I should move all the functions (JS and python) into a `/api` folder at the root of my project, at which point vercel will automatically deploy each function in an appropriate runtime based on its file extension.

However it seems Nextjs itself has a hardcoded assumption that api files are defined at `/pages/api` (or `/src/pages/api`). This means that while my app now works when deployed by Vercel, it breaks on local development. I tried keeping both a `/pages/api` and `/api` folder but it seems that's not recommended either.

Any idea how I can prevent local dev from breaking?

**Comment by DiggidyDev** [comment]
+1 same issue I'm running into at the moment.  I want to have `go` runtime Vercel functions in addition to my app router API routes, but I can't have them both at the same time in local development.

Deploying to Vercel isn't an issue - they both run there, but local development doesn't allow it.

**Comment by dmarcucci** [comment]
How do you run your project locally?

I have a project with both Edge functions and Python endpoints. I run it using the `vercel` command (`vercel dev` to be exact) and not `next dev`. When doing this, I have no issue calling either a Edge function or Python endpoint locally.

**Reply by DiggidyDev** [reply]
I'm using `vercel dev`.

I got around the issue by moving my `/src/app/api` routes to the top-level `/api` folder and refactored the routes to use the `@vercel/node` request types instead.

For clarification, if I had `/api/foo.go` alongside `/src/app/api/bar/route.ts`, the behaviour observed was:
|  | `GET /api/foo` | `GET /api/bar` |
|-----------------|:-----------------:|:-----------------:|
| **`vercel dev`**| ✅ | ❌ |
| **`next dev`**| ❌ | ✅ |
| **Deployed to Vercel**| ✅ | ✅ |

###### P.S. Apologies for the late reply, only just saw this notification :^)

**Reply by dmarcucci** [reply]
Hmm. Strange because I have both Python and JS endpoints located under `/api` and they both work fine with `vercel dev`.

For the `/src/app/api/bar/route.ts` endpoint, you may need to do something like `/api/bar.ts` or `api/bar/index.ts`. I think I have the endpoints in the latter convention and it works no problem.
